language/diem-vm/src/diem_transaction_executor.rs

```diff
@@ -590,6 +591,41 @@ impl DiemVM {
         ))
     }
 
+    fn preload_cache(
+        signature_verified_block: &[PreprocessedTransaction],
+        data_cache: &mut StateViewCache,
+        in_parallel: bool,
+    ) {
+        // generate a collection of addresses
+        let mut addresses_to_preload = Vec::new();
+        for txn in signature_verified_block {
+            if let PreprocessedTransaction::UserTransaction(txn) = txn {
+                if let TransactionPayload::Script(script) = txn.payload() {
+                    addresses_to_preload.push(txn.sender());
+
+                    for arg in script.args() {
+                        if let TransactionArgument::Address(address) = arg {
+                            addresses_to_preload.push(*address);
+                        }
+                    }
+                }
+            }
+        }
+        if in_parallel {
+            rayon::scope(|scope| {
+                scope.spawn(|_| {
+                    for address in addresses_to_preload {
+                        data_cache.get(&AccessPath::new(address, Vec::new())).ok();
+                    }
+                });
+            });
```

Comment:
is there any benefit of using a different thread here? iiuc, the semantic of the `scope` is equivalent to fork-join so it shouldn't be faster than running in the current thread?

Reply:
If it runs in the current thread, won't it block execution of the transactions? I thought we wanted to pipeline the prefetching with the execution....

Reply:
Oh, I get it. You're saying that rayon::scope will block regardless. I checked the docs: "When the closure returns, it will block until all tasks that have been spawned into s complete." That doesn't seem like what we want.

Reply:
yeah, I think it has to block here since we're passing a &mut DataCache here which can't be taken to another thread outlive the function

Reply:
I'm not sure if this would help much but 
```
addresses_to_preload.into_par_iter().map(|addr| data_cache.get(..)).collect()
```
could run all the fetch in parallel

Reply:
@zekun000, are you saying that the execution thread will wait for the prefetching thread to finish?
And, @bob-wilson, are you saying that the prefetching thread should quit (and not continue the prefetching) once the execution thread is done?

The into_par_iter will still await on a joint, right?

Reply:
yes, from what i understand, the scope is essentially a join at the end.

and yes, into_par_iter will join too but it can leverage more threads to execute (the current version only use one, but if the task is trivial it might not have big difference).

Reply:
I was assuming that we would want to run the prefetching in parallel with the execution. I can see why that's a problem with mutating the DataCache, but it seems like it would be important to get the best performance. The ideal would be that only the first transaction has to wait to load resources and by the time each of the remaining transactions starts to execute, the cache is already preloaded -- you would almost totally hide the time to load data. If we have to wait to start executing the first transaction until all the resources for the entire block are prefetched, then it won't be as fast.

Reply:
> run the prefetching in parallel with the execution

this was my impression as well, I imagine it'd be sending an Arc<RwLock<Cache>>> to another thread (not sure if read/write conflict matters here)

Reply:
I didn't realize that the scoped thread will join at the end, that's definitely not what we want. Let me see if I can send the Arc<RwLock>> instead of &mut DataCache as Zekun suggests.

Reply:
The current implementation would work by issueing a number of threads to prefetch the data, join after they all fetch the data and starts the execution, which would avoid cache invalidation problem. If we want to implement things like a separate prefetching thread which fetches data asynchronously, we will need to go one level up: instead of passing a `&mut DataCache`, we should really be reading from the `&dyn StateView`, as that read itself is stateless. However, the internal implementation of the `StateView` that gets passed into the DiemVM, the `VerifiedStateView` has a cache over there that could hold the prefetching result without interfering the subsequent writes.